src/ios/BluetoothLePlugin.m

```diff
@@ -1583,36 +1584,17 @@ - (void)writeQ:(CDVInvokedUrlCommand *)command {
 
   //Set the callback
   [self addCallback:characteristic.UUID forConnection:connection forOperationType:operationWrite forCallback:command.callbackId];
-
+  
   //Get the write type (response or no response)
-  int writeType = [self getWriteType:obj];
-
-  NSUInteger length = [value length];
-  NSUInteger chunkSize = 20;
-  NSUInteger offset = 0;
-  do {
-    NSUInteger thisChunkSize = length - offset > chunkSize ? chunkSize : length - offset;
-    NSData* chunk = [NSData dataWithBytesNoCopy:(char *)[value bytes] + offset length:thisChunkSize freeWhenDone:NO];
-
-    offset += thisChunkSize;
-    [peripheral writeValue:chunk forCharacteristic:characteristic type:writeType];
-  } while (offset < length);
-
-  //Write without response won't execute any callbacks, so return immediately
-  if (writeType == CBCharacteristicWriteWithoutResponse) {
-    NSMutableDictionary* returnObj = [NSMutableDictionary dictionary];
-
-    [self addDevice:peripheral :returnObj];
-    [self addCharacteristic:characteristic :returnObj];
-
-    [self addValue:value toDictionary:returnObj];
-
-    [returnObj setValue:statusWritten forKey:keyStatus];
-
-    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:returnObj];
-    [pluginResult setKeepCallbackAsBool:false];
-    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
-  }
+  writeQtype = [self getWriteType:obj];
```

Comment:
It's been a while since I wrote the old code. Why did so much of this need to change? I thought you'd be able to just update the chunk size and that's it.

Why does noResponse not work for you? I thought that would be quicker.

Reply:
I'm implementing OTA Firmware update in my app and i need to write specific value to a characteristic before and after writing file into device. Since `writeQ` on iOS would return immediatly the app would write OPERATION_END while file was being uploaded.
To address that i used without `type=noResponse` and it took ~110secs to upload the files.
This was too long to wait for and had to be improved.

Reply:
Hmm, still not completely following. What calls are you making to the plugin?

And what's the performance benefit of the current changes over just increasing the chunk size to something larger?

Reply:
I tried using `write` while looping through the files, i sent chunks of 244 bytes for Android and 180 bytes for iOS. On Android i would first call `requestConnectionPriority('high')` and `mtu(250)` and would use `type=noResponse`. On iOS i had to use without `noResponse` since the device would always return error afterwards. Here are the benchmarks for these calls (values are in seconds):

### Using Write

#### Android (without response)

|             | run 1 | run 2 | run 3 |     Avg |
| ----------- | -----: | -----: | -----: | ------: |
| apploader   |  6,466 |  6,042 |  6,091 |    ~6,2 |
| application | 31,019 | 32,786 | 31,910 |  31,905 |
| total       | 37,485 | 38,828 | 38,001 | ~38,105 |

#### iOS (with response)

|             |  run 1 | run 2 | run 3 |      Avg |
| ----------- | ------: | -----: | -----: | -------: |
| apploader   |  16,894 | 17,370 | 18,239 |   17,501 |
| application |  90,541 |   88,8 | 91,921 |   90,421 |
| total       | 107,435 | 106,17 | 110,16 | ~107,922 |

Then i tried `writeQ`, again `type=noResponse` only worked for Android. I recorded only 2 runs for Android, one with 20 byte chunk size and another with 244 byte. Here's the results:

### Using WriteQ

#### Android (without response)

|             | run 1 (20 b/p) | run 2 (244 b/p) | run 3 | Avg |
| ----------- | --------------: | ---------------: | -----: | --: |
| apploader   |           9,666 |            0,805 |        |     |
| application |          50,387 |            9,644 |        |     |
| total       |          60,053 |           10,449 |        |     |

Here we can see that chunk size really plays a important role in the amount of time it takes to upload a file, the difference was substantial.

For iOS i only recorded the results of my implementation of writeQ since it allows for `type=noResponse`, results:

#### iOS (all at 180 b/p)

|             | run 1 | run 2 | run 3 |    Avg |
| ----------- | -----: | -----: | -----: | -----: |
| apploader   |  5,168 |  5,197 |  5,014 |  5,126 |
| application | 29,373 | 29,643 | 29,107 | 29,374 |
| total       | 34,541 | 34,840 | 34,121 | 34,501 |
-----
> And what's the performance benefit of the current changes over just increasing the chunk size to something larger?

One way to do this would be to just increase the value of the chunks in the source code but seeing as i was having issues with chunk size myself where for instance in iOS using 180 byte sized chunk works fine with silabs board but if i increased that value to 182 or 184 for example the device would return an error i thought that would be poor quality code and it would probably only fit my needs and not everyone. Also keep in mind that the MTU value can only be negotiated in Android, iOS does it automatically so if we were to hardcode a chunk value greater than 20 and that be higher than supported by a device then `writeQ` would not be a option.

So i thought the best possible solution was to keep things as they were as much as possible, introduce a optional parameter where the user decides the size of the chunk to be sent. This way the only change to whoever is using this method out there is the fact that in iOS with `type=noResponse` the method will only return after writing the payload and the fact that it only works on iOS11+ which is supported on iPhone5s+.

Thanks for your patience and i hope i've clarified your doubts.

Reply:
Thanks for providing the data!

By changing the chunk size, I meant allow the user to override the default of 20.

What type of error we you receiving on noResponse with a normal #write() on iOS? My AngularJS based examples using  noResponse for the "firmware" upgrade example: https://github.com/randdusing/ng-cordova-bluetoothle/blob/master/example/index.js#L626

Although it looks like iOS11 changed write without response significantly, which is maybe why this doesn't work anymore. I read up on canSendWriteWithoutResponse and peripheralIsReadyToSendWriteWithoutResponse. Does your code need to consider canSendWriteWithoutResponse?

I should break out my iOS device...

Reply:
>What type of error we you receiving on noResponse with a normal #write() on iOS?

The error was not specific to the `write` operation, it's just that i need to know when the operation ends so i can then write `0x03` to the device to signal end of upload and so this is only possible **without** `type=noResponse`.

>Does your code need to consider canSendWriteWithoutResponse?

There's very little documentation on this and what i could find about it in the forums indicated that this was very problematic ([source](https://developer.apple.com/forums/thread/80376)). I've managed to find a sample from apple on how to transfer data between low energy devices using `WriteWithoutResponse` and in it they use both ([sample](https://developer.apple.com/documentation/corebluetooth/transferring_data_between_bluetooth_low_energy_devices)).

I will add in the `canSendWriteWithoutResponse` into the changes based on the sample.

Reply:
@randdusing i've added `canSendWriteWithoutResponse`. I tested this on iPhone7 with latest iOS and works. But now there's a possible issue, if `canSendWriteWithoutResponse` returns negative i'm not sure what the end result will be.

In their documentation Apple says:
>If this value is NO, flushing all current writes sets the value to YES. This also results in a call to the delegate’s peripheralIsReadyToSendWriteWithoutResponse

But i have no clue how to flush all current writes. Also in the sample i linked there's no reference to such thing either.
Please let me know what else can be done here.

Reply:
If it returns NO, I think you just need to wait for peripheralIsReadyToSendWriteWithoutResponse to return and then continue writing. I'm thinking you want to do something like this:

- Check canSendWriteWithoutResponse
- If YES, write immediately
- If NO, wait for peripheralIsReadyToSendWriteWithoutResponse and then write 
- Repeat first step

Reply:
Alright, done!
It's just like in the sample, write while `canSendWriteWithoutResponse = YES`. If we get `NO` upload will resume once callback kicks in.

Reply:
Hey @randdusing 

is there anything else i can do to improve this PR?

Hope to hear from you soon. Thanks

Reply:
Sorry for the delay. I was thinking more about safety than performance with canSendWriteWithoutResponse, but hard to know the effects with Apple's sparse documentation.

So just to clarify for semver purposes, do I need to update the major version? If it only affects older versions of iOS, I'm not too concerned. But if there's another change in behavior, I probably need to update.

Reply:
With the changes i've made whoever already uses this plugin will have no issue if using iOS10+ otherwise it might fail (can't verify this).
Also for those using `writeQ` they will now get a callback on iOS with `type=noResponse` but this to me does not represent a breaking change. Everything else is additive (optional to the user) changes.

If this is reason to release a new major version of the plugin is up to you.

Reply:
Hey @randdusing sorry to keep bugging you but do you think a release will be possible until next monday? I have an internal release scheduled for next monday and it would be nice to include these changes.

Thanks.

Reply:
Merged. I made this a major version just in case.